server/pkg/scene/sceneops/plugin_migrator.go

```diff
@@ -119,105 +95,76 @@ func (s *PluginMigrator) MigratePlugins(ctx context.Context, sc *scene.Scene, ol
 		}
 	}
 
-	// レイヤー
+	// layers
 	for _, l := range layers {
 		if l == nil {
 			continue
 		}
 		ll := *l
-		llp := ll.Plugin()
-		lle := ll.Extension()
-
-		// 不正なレイヤーの検出
-		if llp != nil && lle != nil && (*llp).Equal(oldPluginID) {
-			if newPlugin.Extension(*lle) == nil {
-				// 削除
-				removedLayers = append(removedLayers, ll.ID())
-				if p := ll.Property(); p != nil {
-					removedPropertyIDs = append(removedPropertyIDs, *p)
-				}
-				if ib := ll.Infobox(); ib != nil {
-					removedPropertyIDs = append(removedPropertyIDs, ib.Property())
-					for _, f := range ib.Fields() {
-						removedPropertyIDs = append(removedPropertyIDs, f.Property())
-					}
-				}
-				continue
-			}
+		llp, lle := ll.Plugin(), ll.Extension()
+
+		// Detect invalid layers
+		if isInvalidLayer(llp, lle, oldPlugin, newPlugin) {
+			removedLayers.AppendLayers(ll.ID())
+			removedPropertyIDs = append(removedPropertyIDs, collectLayerPropertyIds(ll)...)
+			continue
 		}
 
 		if p := ll.Property(); p != nil {
 			propertyIDs = append(propertyIDs, *p)
 		}
 
-		// 不正なInfoboxFieldの削除
-		if ib := ll.Infobox(); ib != nil {
-			removeFields := []layer.InfoboxFieldID{}
-			for _, f := range ib.Fields() {
-				if newPlugin.Extension(f.Extension()) == nil {
-					removeFields = append(removeFields, f.ID())
-					removedPropertyIDs = append(removedPropertyIDs, f.Property())
-				} else {
-					propertyIDs = append(propertyIDs, f.Property())
-				}
+		if ll.Infobox() == nil {
+			continue
+		}
+
+		// Remove invalid Infobox Fields
+		modified := false
+		for _, f := range ll.Infobox().Fields() {
+			if f.Plugin() != oldPlugin.ID() {
+				continue
 			}
-			for _, f := range removeFields {
-				ib.Remove(f)
+			modified = true
+			if newPlugin.Extension(f.Extension()) == nil {
+				ll.Infobox().Remove(f.ID())
+				removedPropertyIDs = append(removedPropertyIDs, f.Property())
+			} else {
+				propertyIDs = append(propertyIDs, f.Property())
 			}
```

Comment:
@rot1024, should not we update the Infobox field plugin as well?

Reply:
infobox block's properties should be migrated